crates/runtime/src/backtrace.rs

```diff
@@ -0,0 +1,251 @@
+        mut f: impl FnMut(Frame) -> ControlFlow<()>,
+    ) -> ControlFlow<()> {
+        log::trace!("=== Tracing through contiguous sequence of Wasm frames ===");
+        log::trace!("first_wasm_sp = 0x{:016x}", first_wasm_sp);
+        log::trace!("   initial pc = 0x{:016x}", pc);
+        log::trace!("   initial fp = 0x{:016x}", fp);
+
+        // In our host-to-Wasm trampoline, we save `-1` as a sentinal SP
+        // value for when the callee is not actually a core Wasm
+        // function (as determined by looking at the callee `vmctx`). If
+        // we encounter `-1`, this is an empty sequence of Wasm frames
+        // where a host called a host function so the following
+        // happened:
+        //
+        // * We entered the host-to-wasm-trampoline, saved (an invalid
+        //   sentinal for) entry SP, and tail called to the "Wasm"
+        //   callee,
+        //
+        // * entered the Wasm-to-host trampoline, saved the exit FP and
+        //   PC, and tail called to the host callee,
+        //
+        // * and are now in host code.
+        //
+        // Ultimately, this means that there are 0 Wasm frames in this
+        // contiguous sequence of Wasm frames, and we have nothing to
+        // walk through here.
+        if first_wasm_sp == -1_isize as usize {
+            log::trace!("Empty sequence of Wasm frames");
+            return ControlFlow::Continue(());
+        }
+
+        // We use `0` as a sentinal value for when there is not any Wasm
+        // on the stack and these values are non-existant. If we
+        // actually entered Wasm (see above guard for `-1`) then, then
+        // by the time we got here we should have either exited Wasm
+        // through the Wasm-to-host trampoline and properly set these
+        // values, or we should have caught a trap in a signal handler
+        // and also properly recovered these values in that case.
+        assert_ne!(pc, 0);
+        assert_ne!(fp, 0);
+        assert_ne!(first_wasm_sp, 0);
+
+        // The stack grows down, and therefore any frame pointer we are
+        // dealing with should be less than the stack pointer on entry
+        // to Wasm.
+        assert!(first_wasm_sp >= fp, "{first_wasm_sp:#x} >= {fp:#x}");
+
+        // The stack pointer should always be aligned to 16 bytes
+        // *except* inside function prologues where the return PC is
+        // pushed to the stack but before the old frame pointer has been
+        // saved to the stack via `push rbp`. And this happens to be
+        // exactly where we are inside of our host-to-Wasm trampoline
+        // that records the value of SP when we first enter
+        // Wasm. Therefore, the SP should *always* be 8-byte aligned but
+        // *never* 16-byte aligned.
+        if cfg!(target_arch = "x86_64") {
+            assert_eq!(first_wasm_sp % 8, 0);
+            assert_eq!(first_wasm_sp % 16, 8);
+        } else if cfg!(target_arch = "aarch64") {
```

Comment:
Perhaps actually remove the condition here so other platforms can add conditions as appropriate? (if they don't already fall back to this case)